### PR TITLE
extend+improve HIR types documentation

### PR DESCRIPTION
**Pull request by llogiq**

### CODE REVIEW

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -919,13 +919,13 @@ pub enum Expr_ {
     /// Inline assembly (from `asm!`), with its outputs and inputs.
     ExprInlineAsm(InlineAsm, Vec<P<Expr>>, Vec<P<Expr>>),
 
-    /// A struct literal expression.
+    /// A struct or enum variant literal expression.
```
Comment (eddyb) [inline src/librustc/hir/mod.rs]: I think "struct-like variant" is the correct term here.

File: src/librustc/hir/mod.rs [hunk 2]
```diff
@@ -950,14 +950,21 @@ pub struct QSelf {
...
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug, Copy)]
 pub enum MatchSource {
+    /// A `match _ { .. }`
     Normal,
+    /// An `if let _ = _ { .. }` (optionally with `else { .. }`
     IfLetDesugar {
         contains_else_clause: bool,
     },
+    /// A `while let _ = _ { .. }` (which was desugared to a
+    /// `loop { match _ { .. } }`
```
Comment (eddyb) [inline src/librustc/hir/mod.rs]: Paren isn't closed.

**Comment by llogiq** [comment]
@eddyb thank you, done.

**Pull request merged**